inc/parser/xhtml.php

```diff
@@ -1607,6 +1607,53 @@ public function _formatLink($link) {
         return $ret;
     }
 
+    /**
+     * Returns effective image width and height based on the specified desired
+     * width/height, if any, and the meta data loaded from the image file.
+     * Including the effective image sizes in the HTML increases the stability
+     * of content, i.e. results in a better Cumulative Layout Shift (CLS) score:
+     * https://web.dev/cls/
+     * https://web.dev/optimize-cls/#images-without-dimensions
+     *
+     * @author Michael Stapelberg
+     * @param  int      $desiredWidth     specified width or null
+     * @param  int      $desiredHeight    specified height or null
+     * @param  JpegMeta $jpeg             meta data loaded from the image
+     * @return array of int               effective width and height
+     */
+    private function _image_sizes($desiredWidth, $desiredHeight, $jpeg) {
+        if($jpeg === false) {
+            return array($desiredWidth, $desiredHeight);
+        }
+
+        // If no dimensions are specified, take the dimensions from the file
+        if(is_null($desiredWidth) && is_null($desiredHeight)) {
+            return array($jpeg->getWidth(), $jpeg->getHeight());
+        }
+
+        if(is_null($desiredWidth) && !is_null($desiredHeight)) {
+            // height specified, calculate effective width:
+            $ratio = $desiredHeight / $jpeg->getHeight();
+            $width = floor($jpeg->getWidth() * $ratio);
+            return array($width, $desiredHeight);
+        }
+
+        if(!is_null($desiredWidth) && is_null($desiredHeight)) {
+            // width specified, calculate effective height:
+            $ratio = $desiredWidth / $jpeg->getWidth();
+            $height = floor($jpeg->getHeight() * $ratio);
+            return array($desiredWidth, $height);
+        }
+
+        // Both dimensions fully specified:
+        return array($desiredWidth, $desiredHeight);
+    }
+
+    private function _load_jpeg_meta($src) {
+        require_once(DOKU_INC.'inc/JpegMeta.php');
```

Comment:
Is this line needed. The autoloader of DokuWiki should arrange this already. Please remove.

Reply:
Done.